src/xdg.c

```diff
@@ -320,6 +320,111 @@ xdg_toplevel_view_minimize(struct view *view, bool minimized)
 	/* noop */
 }
 
+static struct wlr_xdg_toplevel *
+top_parent_of(struct view *view)
+{
+	struct wlr_xdg_toplevel *toplevel = xdg_toplevel_from_view(view);
+	while (toplevel->parent) {
+		toplevel = toplevel->parent;
+	}
+	return toplevel;
+}
+
+enum z_direction {
+	LAB_TO_FRONT,
+	LAB_TO_BACK,
+};
+
+static void
+move_sub_views(struct view *parent, enum z_direction z_direction)
+{
+	assert(parent);
+
+	if (parent->type != LAB_XDG_SHELL_VIEW) {
+		return;
+	}
+
+	struct wlr_xdg_toplevel *parent_toplevel = xdg_toplevel_from_view(parent);
+	struct view *view, *next;
+	wl_list_for_each_reverse_safe(view, next, &parent->server->views, link)
+	{
```

Comment:
This seems like a copy of `xdg_toplevel_view_get_children()`.
Is it worth combining the two?

Pros:
- less code duplication

Cons:
- additional heap allocations by `wl_array`.

Reply:
No. `move_sub_views()` is quite subtle in that it uses the `_safe` version of the `for_each` macro because at the bottom of the loop we change the stacking order!

I think there are other areas of the code which could benefit from this type of approach though :smile:
When I get a minute I will experiment with `osd.c` to see if we could get the views we want to iterate over as an array (possibly with options to include/exclude different things). The advantage of wl_array over wl_list in this type of short-lived situation is that we wouldn't need any extra `wl_list link_foo` entries in the structs.

Reply:
But wouldn't the result be the same if we do something like this?
```c
static void
move_sub_views(struct view *parent, enum z_direction z_direction)
	assert(parent);

	if (parent->type != LAB_XDG_SHELL_VIEW) {
		return;
	}

	struct wl_array subviews;
	wl_array_init(&subviews);
	xdg_toplevel_view_get_children(parent, &subviews);

	struct view *view;
	wl_array_for_each(view, &subviews) {
		if (z_direction == LAB_TO_FRONT) {
			view_impl_move_to_front(view);
		} else if (z_direction == LAB_TO_BACK) {
			view_impl_move_to_back(view);
		}
	}
	wl_array_release(&subiews);
}
```

Reply:
This is a very good point. Should we do this in a separate PR?
Either way, I'm too tired tonight.